runners/spark/src/main/java/org/apache/beam/runners/spark/SparkRunner.java

```diff
@@ -216,6 +219,34 @@ public CompositeBehavior enterCompositeTransform(TransformTreeNode node) {
       return CompositeBehavior.ENTER_TRANSFORM;
     }
 
+    private boolean shouldDiffer(TransformTreeNode node) {
```

Comment:
I agree with what you are saying - mandatory translations are quite different from optional composite translations. By definition, all translation done in `enterCompositeTransform` is optional, and all translation in `visitPrimitiveTransform` is mandatory. The translators in the `SparkRunner` are split pretty evenly between these two.

I'm not saying they should be separate classes - they do basically the same thing. But the primitive translators don't have the question "should I be applied to this transform?" and all the optional translators _do_ have that question. Another way of saying what I mean is that the `if (Combine.PerKey) { ... } else if (Combine.Globally) { ... }` looks like it should be dynamic dispatch instead of conditionals.

I'm going to say LGTM because so far there is just the one exception. If other translations become conditional, then I definitely think you should break them out into classes that encapsulate the knowledge of when they apply.

By the way, I think you might mean `defer` instead of `differ`? Or else I don't quite know what the name of the method is implying.